File: src/com/tyrcho/gui/toolkit/RadioButtonGroup.java
package com.tyrcho.gui.toolkit;

import java.awt.BorderLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.FocusListener;
import java.security.InvalidParameterException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import javax.swing.ButtonGroup;
import javax.swing.JRadioButton;
import javax.swing.SwingConstants;

import com.tyrcho.gui.field.IInputField;


/**
 * Displays a group of Radio Buttons and allows the selection to be 
 * retrieved as an object.
 * @author MDA.
 */
public class RadioButtonGroup 
extends Panel
implements IInputField, SwingConstants
{
	private Object selectedValue;
	private Map buttons=new HashMap();
	private Map values;
	private ButtonGroup buttonGroup=new ButtonGroup();
	//Used to deselect other buttons.
	private JRadioButton deselect=new JRadioButton();
	
	/**
	 * @param orientation either VERTICAL or HORIZONTAL, used to layout buttons
	 * @param values a map with keys=labels, values=the object to return
	 */
	public RadioButtonGroup(int orientation, Map values)
	{
		super(new BorderLayout());
		this.values=values;
		javax.swing.Box buttonsBox;
		switch (orientation)
		{
			case VERTICAL : buttonsBox=Box.createVerticalBox();
			case HORIZONTAL : buttonsBox=Box.createHorizontalBox();
			break;
			default : throw new InvalidParameterException(orientation +" is not a valid orientation.");
		}
		for (Iterator i=values.keySet().iterator(); i.hasNext(); )
		{
			String label=(String)i.next();
			JRadioButton button=createRadioButton(label);
			final Object value=values.get(label);
			buttons.put(label, button);
			button.addActionListener(new ActionListener()
			{
				public void actionPerformed(ActionEvent e)
				{
					selectedValue=value;
				}
			});
			buttonsBox.add(button);
			if (i.hasNext())
			{
				buttonsBox.add(Box.createGlue());
			}
			buttonGroup.add(button);
		}
		buttonGroup.add(deselect);
		add(buttonsBox, BorderLayout.CENTER);
	}
	
	public void addFocusListener(FocusListener l)
	{
		if (buttons!=null)
		{
			for (Iterator i=buttons.values().iterator(); i.hasNext(); )
			{
				JRadioButton button=(JRadioButton) i.next();
				button.addFocusListener(l);
			}
		}
	}
	
	public void removeFocusListener(FocusListener l)
	{
		if (buttons!=null)
		{
			for (Iterator i=buttons.values().iterator(); i.hasNext(); )
			{
				JRadioButton button=(JRadioButton) i.next();
				button.removeFocusListener(l);
			}
		}
	}

	public void addActionListener(ActionListener l)	
	{
		if (buttons!=null)
		{
			for (Iterator i=buttons.values().iterator(); i.hasNext(); )
			{
				JRadioButton button=(JRadioButton) i.next();
				button.addActionListener(l);
			}
		}
	}
	
	public void removeActionListener(ActionListener l)	
	{
		if (buttons!=null)
		{
			for (Iterator i=buttons.values().iterator(); i.hasNext(); )
			{
				JRadioButton button=(JRadioButton) i.next();
				button.removeActionListener(l);
			}
		}
	}
	
	/**
	 * Can be override to return a specific implementation of JRadioButton.
	 */
	protected JRadioButton createRadioButton(String label)
	{
		return new JRadioButton(label);
	}
	
	public Object getCurrentValue()
	{
		return selectedValue;
	}
	
	public void setCurrentValue(Object newValue)
	{
		boolean found=false;
		for (Iterator i=values.keySet().iterator(); i.hasNext(); )
		{
			Object label=i.next();
			Object value=values.get(label);
			if (value.equals(newValue))
			{
                found=true;
				JRadioButton button=(JRadioButton)buttons.get(label);
				button.doClick();
			}
		}	
		if (!found)
		{
			deselect.setSelected(true);
		}	
	}

	public void clear()
	{
		setCurrentValue(null);
	}
	
	public void setEditable(boolean editable)
	{
		for (Iterator i=buttons.values().iterator(); i.hasNext(); )
		{
			JRadioButton button=(JRadioButton)i.next();
			button.setEnabled(editable);
		}
	}
}

File: src/info/daviot/gui/RadioButtonGroup.scala
package info.daviot.gui

import info.daviot.gui.field.IInputField;

import java.awt.BorderLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.FocusListener;
import java.security.InvalidParameterException;

import javax.swing.Box;
import javax.swing.ButtonGroup;
import javax.swing.JPanel;
import javax.swing.JRadioButton;
import javax.swing.SwingConstants;

/**
 * Displays a group of Radio Buttons and allows the selection to be retrieved as
 * an object.
 *
 * @author mdaviot
 */
class RadioButtonGroup(orientation: Int, values: Map[String, Any]) extends JPanel(new BorderLayout()) with IInputField {
  var selectedValue: Any = _
  val buttonGroup = new ButtonGroup();
  // Used to deselect other buttons.
  val deselect = new JRadioButton();

  val buttonsBox = if (SwingConstants.VERTICAL == orientation) Box.createVerticalBox() else Box.createHorizontalBox()
  val buttons = values.map(kv => (kv._2 -> new JRadioButton(kv._1)))
  buttons foreach (kv => {
    kv._2.addActionListener(new ActionListener() { def actionPerformed(e: ActionEvent) { selectedValue = kv._1 } })
    buttonsBox.add(kv._2)
    buttonsBox.add(Box.createGlue())
    buttonGroup.add(kv._2);
  })

  buttonGroup.add(deselect);
  add(buttonsBox, BorderLayout.CENTER);

  override def addFocusListener(l: FocusListener) { buttons.values.foreach(_.addFocusListener(l)) }
  override def removeFocusListener(l: FocusListener) { buttons.values.foreach(_.removeFocusListener(l)) }

  def addActionListener(l: ActionListener) { buttons.values.foreach(_.addActionListener(l)) }
  def removeActionListener(l: ActionListener) { buttons.values.foreach(_.removeActionListener(l)) }

  def setEditable(editable: Boolean) { buttons.values.foreach(_.setEnabled(editable)) }

  def getCurrentValue = selectedValue.asInstanceOf[AnyRef]

  def setCurrentValue(newValue: Any) {
    buttons.find(_._1 == newValue) match {
      case Some(b) => b._2.doClick
      case None => deselect.setSelected(true)
    }
  }

  def clear() { setCurrentValue(null); }

}
